File: Leetcode/023_Merge_k_Sorted_Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    struct Node {
      int val;
      int from;
      Node(int x, int y) : val(x), from(y) {}
    };
    
    struct NodeCmp{
        bool operator()(const Node &a, const Node &b){
            return a.val > b.val;
        }
    };

    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<Node, vector<Node>, NodeCmp> heap;
        ListNode *head = new ListNode(0);
        ListNode *tail = head;
        
        for (int i = 0; i < lists.size(); i ++) {
            if (lists[i]) {
                heap.push(Node(lists[i]->val, i));
            }
        }
        
        while (!heap.empty()) {
            Node tmp = heap.top();
            heap.pop();
            ListNode *tmpL = new ListNode(tmp.val);
            tail->next = lists[tmp.from];
            tail = tail->next;
            lists[tmp.from] = lists[tmp.from]->next;
            if (lists[tmp.from]) {
                heap.push(Node(lists[tmp.from]->val, tmp.from));
            }
        }
        
        return head->next;
    }
};

File: Leetcode/023_Merge_k_Sorted_Lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> pq = new PriorityQueue<ListNode>(
            new Comparator<ListNode>() {
                @Override
                public int compare(ListNode a, ListNode b) {
                    return a.val - b.val;
                }
            }
        );
        for (ListNode list : lists) 
            if (list != null)
                pq.add(list);
        
        ListNode head = new ListNode(0);
        ListNode tail = head;
        while (!pq.isEmpty()) {
            tail.next = pq.poll();
            tail = tail.next;
            if (tail.next != null)
                pq.add(tail.next);
        }
        return head.next;
    }
}

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> pq = new PriorityQueue<>(new Comparator<ListNode>(){
            public int compare(ListNode a, ListNode b) {
                return a.val - b.val;
            }
        });
        for (ListNode list : lists) 
            if (list != null) pq.add(list);
        ListNode head = new ListNode(0);
        ListNode tail = head;
        while (!pq.isEmpty()) {
            tail.next = pq.poll();
            tail = tail.next;
            if (tail.next != null)
                pq.add(tail.next);
        }
        return head.next;
    }
}